torch/testing/_internal/common_utils.py

```diff
@@ -885,7 +885,8 @@ def safeCoalesce(self, t):
     # in https://github.com/pytorch/pytorch/pull/32538.
     # dtype name : (rtol, atol)
     dtype_precisions = {
-        torch.float16    : (0.001, 1e-5),
+
+        torch.float16    : (0.005, 1e-4),
```

Comment:
@zou3519 FYI I made a small adjustment to tolerances here in this PR (post-approval) based on the idea that a half type holds about 3 decimal digits. This rtol might still be a bit too tight of a default for many half cases, but we can always relax it more later if we find ourselves needing to tweak too often.
I think it makes sense for bfloat to have a tighter default atol than half since it has higher dynamic range (the same exponent bits as float32).

Reply:
Since half type holds more than 3 decimal places, 0.001 seems like a good number for `rtol`. Were there specific cases where a test needs more than 0.001 for `rtol`?

Note that it's not that "a half type holds about 3 decimal digits", to be more exact, a half type holds "more than 3 decimal digits": https://en.wikipedia.org/wiki/Half-precision_floating-point_format

Reply:
If there's anything that needed something less tight than `(0.001, 1e-5)`, then using the `half_tol `override mechanism would preserve the old behavior without relaxing all of the tolerances

Reply:
> Since half type holds more than 3 decimal places

I don't think that's correct, "safely holds" doesn't mean it will retain that precision after an operation. We don't check tolerances of 1e-7 on float, nor 1e-15 on double, because so many of our operators are non-trivial matrix operations and we don't want to have to manually set tolerances on every test. "Safely holds 3" is why I wanted to widen it and why even 0.005 seems pretty tight.

E.g. testing a simple sum operation on moderately sized half tensor already brings us to a tolerance of ~0.004:
```
r=torch.arange(0, 99, dtype=torch.float)
r.sum().item()
Out[63]: 4851.0
r.to(torch.half).sum().item() / r.sum().item() - 1
Out[64]: -0.003916718202432445
```
Requiring tests to special-case the tolerance on tests this simple feels like it creates excessive overhead and complication across the board.

> using the half_tol override mechanism

My motivation for loosening it wasn't specifically for these tests, but because I think it's a more reasonable default. It would be easy to adapt these tests to any default tolerance, I just don't feel like a tenth of a percent of tolerance makes sense when checking results of operations on a dtype that doesn't precisely represent `3333`.

Reply:
`torch.sum` on a CPU half tensor is not the best example because we don't expect users to use CPU half. We don't do dtype accumulation on CPU half tensors for torch.sum, but we do accumulate into float for CUDA half tensors, so the precision on torch.sum for CPU half tensors is "bad".

The reason why I am wary about relaxing precisions is it's possible that in the future we regress in precision (an operation that was able to support 0.001 rtol before doesn't anymore) and this causes some pre-existing model to not converge anymore. It would be difficult to bisect to such a commit.

That's the conservative side in me speaking and I am not sure how realistic that situation is. But I've spent a good chunk of time in the past debugging why models don't converge anymore on new releases of PyTorch. My personal opinion is that I am not comfortable with relaxing pre-existing half precision testing across the board.

Reply:
> relaxing pre-existing half precision

I wouldn't exactly think of it as pre-existing. I do want to choose _initial_ defaults that make sense for PyTorch broadly. This is a new(ish) parameter, and when I added it I felt like it was easier to be conservative and relax later than to start loose and tighten later. Many (most?) half tolerances are currently being set manually through atol, which feels unsustainable to me.

If we can't pick a tolerance that fits a super-majority of operators then I suspect we're doing something wrong.

> so the precision on torch.sum for CPU half tensors is "bad".

That's interesting, I had tried it on a non-cuda host. If we can reasonably expect most overriding operators fit a tighter tolerance than they do now, maybe there should be a project to fix "bad" operators? I suspect that some of the overrides even outside this looser 0.005 tolerance are also "bad" and should be fixed (and was hoping we'd revisit some of these once this lands and they stand out better).

I think I'll feel more comfortable setting a tighter tolerance if we can feel comfortable asserting that operators outside that tolerance are exceptional/bad in some way. That seems like a hard argument to make when a substantial portion of our tests are already exceptions.

A standard for tolerance should enforce "correctness" under some reasonable default definition of correctness. When most operators are tested merely for whatever worked at the time it was written, then we're only testing for regressions and not that it's not "bad."

A common rarely-overridden default provides a stronger incentive not to land "bad" operators in the first place.

Reply:
I see, I didn't realize how recent the rtol default was. You've also changed the atol default though (from 1e-5 to 1e-4), does that relax any pre-existing tolerances, and if so, by how much? If it turns out that we haven't relaxed it by very much then my concern about regressions becomes moot.

Reply:
The original default self.precision was set to 1e-5 and applied as atol, but I don't think it was successfully used much with half-precision tests. The absolute difference between 1/3 stored in a half and 1/3 as a double is greater than 1e-5, but less than 1e-4, even without performing any further operations.

I am somewhat swayed by the argument that cpu-sum is "bad," if special-handling code for `sum` and similar operators is highlighting a real problem.

Re-running this batch of tests with default rtol set back to 0.001, the test failures I see that we would instead need to explicitly call out with an override are TestTensorDeviceOpsCUDA.:
test_addbmm_cuda_float16
test_addbmm_inplace__cuda_float16
test_addbmm_scalar_cuda_float16
test_addbmm_inplace_scalar_cuda_float16
test_addcdiv_inplace_scalar_cuda_float16
test_addcdiv_scalar_cuda_float16
test_cumsum_cuda_float16
test_cumsum_neg_dim_cuda_float16
test_prod_cuda_float16

If we suspect addbmm, addcdiv, cumsum, and prod have differences due to normal/unimportant/minor precision variations then I'd prefer to continue with 0.005. If we think these are issues like cpu-sum that we can consider "bad" that we could/should "fix," then I'm pretty amenable to lowering back to 0.001. What do you think?

Reply:
@nairbv my concern is only about any potential relaxing of the half precision (i.e., if we were testing something with tighter tolerances before than we are after this PR). You've addressed that concern so I don't have more concerns.

Regarding the default, I don't know "what a good default" should be. Looking at the list of ops above, cumsum doesn't actually accumulate into float (unlike a lot of our other ops, but it's been like that for a few years. Maybe some of the folks on https://github.com/pytorch/pytorch/pull/36458 would know more). `prod` accumulates into `float` on half types, so I'm not sure why it isn't passing rtol=0.001 (https://github.com/pytorch/pytorch/blob/0ff1aa90589ed113ee127dea484f77f3a014efc4/aten/src/ATen/native/cuda/ReduceSumProdKernel.cu#L46). `addbmm` should call straight into the CUDA API so we should at least be able to say that that one is "implemented correctly"